### PR TITLE
Gh159 actions unit testing

### PR DESCRIPTION
**Pull request by SeyfSV**
Hi @dsuch,

Please take a look on this changes. Are they needed for project, especially testing workflows?

Testing are not completely redesigned and does not cover all things (7 tests skipped), but at this moment all of them passing or skipping, and can be used for project testing - pytest pass without error.

### CODE REVIEW

**Comment by dsuch** [comment]
Hi @SeyfSV - can you briefly tell me what the workflows are about or perhaps link me to an overview of the feature? It sounds intriguing and I would like to learn more. Thanks.

**Comment by SeyfSV** [comment]
Overview: https://github.com/features/actions
Workflow: https://help.github.com/en/actions/configuring-and-managing-workflows/configuring-a-workflow

I added some notes to #159 about what was done

**Comment by AlexandreYang** [comment]
👋, Any update on the status of this PR ? I would be glad to land a hand on this to have a CI that test for regressions.

**Comment by SeyfSV** [comment]
Yes, https://github.com/dsuch/pymqi/issues/207#issuecomment-645401720

**Comment by SeyfSV** [comment]
At this moment I merging it with current master

**Comment by SeyfSV** [comment]
I think this ready for merge
@AlexandreYang, can you review it ?

File: .github/workflows/install-tests.yml [hunk 1]
```diff
@@ -0,0 +1,44 @@
+name: install-tests
+on: 
+  push:
+    branches:
+      - master
+      - gh159-actions-unit-testing
```
Comment (AlexandreYang) [inline .github/workflows/install-tests.yml:6]: TODO: remove `gh159-actions-unit-testing` before merging to master

File: .github/workflows/install-tests.yml [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    strategy:
+      matrix:
+        environment: ['macos-latest', 'windows-latest', 'ubuntu-latest']
+        python-version: [2.7, 3.5, 3.6, 3.7, 3.8]
+        mq-client-version: [9.1.4.0]
+        exclude:
+          - environment: windows-latest
+            python-version: 2.7
+          - environment: windows-latest
+            python-version: 3.5
```
Comment (AlexandreYang) [inline .github/workflows/install-tests.yml:20]: What's the reason for excluding those 2 envs?

File: .github/workflows/unit-tests.yml [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+      - name: Cache MQ Client
+        uses: actions/cache@v1
+        with:
+          path: ${{ github.workspace }}/setup-mqclient
+          key: mqclient-${{ runner.os }}-${{ matrix.mq-client-version }}
+
+      - name: Install MQ Client
+        id: setup-mqclient
+        uses: SeyfSV/setup-mqclient@v0.1.3
```
Comment (AlexandreYang) [inline .github/workflows/unit-tests.yml:58]: Nice for `SeyfSV/setup-mqclient` 👍

File: .github/workflows/unit-tests.yml [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+jobs:
+  pymqi_test_job:
+    strategy:
+      max-parallel: 3
+      matrix:
+        environment: ['ubuntu-latest']
+        # windows-latest: runner does allow to run ibmcom/mq linux container
+        # macos-latest: runner does not have docker
+        # macos-latest: has some issues in SeyfSV/setup-mqclient@master action
+        #environment: ['ubuntu-latest', 'macos-latest', 'windows-latest']
```
Comment (AlexandreYang) [inline .github/workflows/unit-tests.yml:18]: Need some clean up ? Any reason not running on platforms other than `ubuntu`?

File: .github/workflows/unit-tests.yml [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        with:
+          name: mq-server-logs
+          path: ./errors
+
+      - name: Upload MQ Client logs
+        if: failure()
+        uses: actions/upload-artifact@v1
+        with:
+          name: mq-client-logs
+          path: ${{ steps.setup-mqclient.outputs.mq-file-path }}/errors
```
Comment (AlexandreYang) [inline .github/workflows/unit-tests.yml:120]: Where does `mq-file-path` come from ?

File: code/README [hunk 6]
```diff
@@ -1,4 +1,7 @@
 
+![unit-tests](https://github.com/dsuch/pymqi/workflows/unit-tests/badge.svg)
```
Comment (AlexandreYang) [inline code/README:2]: Shall we rename `code/README` to `code/README.md` ?

File: code/tests/README [hunk 7]
```diff
@@ -5,26 +5,46 @@ How to run PyMQI's test suite
...
-* Install nose and testfixtures
-  $ sudo pip install nose
-  $ sudo pip install testfixtures
+* Install tox
+  ```
+  $ sudo pip install tox
+  ```
 
-* Set a proper configuration by modifying the defaults in config.py or override
+* Set a proper configuration by modifying the defaults in config.py or override
```
Comment (AlexandreYang) [inline code/tests/README:13]:
```suggestion
* Set a proper configuration by modifying the defaults in config.py or override
```

File: code/tests/test_rfh2.py [hunk 8]
```diff
@@ -219,7 +220,7 @@ def test_buffer_too_short_for_default_rfh2_exception(self):
 
         rfh2 = pymqi.RFH2()
         try:
-            rfh2.unpack(self.single_rfh2_message[0:32])
+            rfh2.unpack(self.single_rfh2_message[0:36])
```
Comment (AlexandreYang) [inline code/tests/test_rfh2.py:223]: What led to this change?

File: code/tests/test_pcf.py [hunk 9]
```diff
@@ -274,6 +273,7 @@ def test_unpack_group(self):
         self.assertEqual(item[pymqi.CMQC.MQCA_Q_NAME].strip(), b'SYSTEM.ADMIN.COMMAND.QUEUE')
         self.assertEqual(item[pymqi.CMQCFC.MQIAMO_PUTS], [14, 0])
 
+    @skip('Python pre 3.7 issues')
```
Comment (AlexandreYang) [inline code/tests/test_pcf.py:276]: Could you add a comment with more details why this is skipped ?

Shall we skip for all python versions or only some ?

File: code/tests/test_pcf.py [hunk 10]
```diff
@@ -23,7 +23,6 @@ def setUpClass(cls):
 
         # max length of queue names is 48 characters
         cls.queue_name = "{prefix}PCF.QUEUE".format(prefix=cls.prefix)
-        cls.pcf = pymqi.PCFExecute(cls.qmgr, response_wait_interval=600)
```
Comment (AlexandreYang) [inline code/tests/test_pcf.py:26]: Shall we remove `cls.pcf.disconnect()` in `tearDownClass` too ? If yes, the whole `tearDownClass` can be removed.

File: code/tests/test_queue_manager.py [hunk 11]
```diff
@@ -151,50 +138,5 @@ def test_inquire(self):
         self.assertEqual(len(attribute_value), pymqi.CMQC.MQ_Q_MGR_NAME_LENGTH)
         self.assertEqual(attribute_value.strip(), expected_value)
 
-    def test_is_connected(self):
```
Comment (AlexandreYang) [inline code/tests/test_queue_manager.py:154]: Why this test is not needed ?

File: tox.ini [hunk 12]
```diff
@@ -24,3 +24,42 @@ max-line-length=160
 # W291 trailing whitespace
 # W293 blank line contains whitespace
 ignore=E112,E121,E122,E123,E124,E125,E126,E127,E128,E129,E131,E225,E226,E231,E251,E261,E302,E401,E713,F403,W291,W293
+
+; [tox]
```
Comment (AlexandreYang) [inline tox.ini:28]:
```suggestion
```

File: .github/workflows/install-tests.yml (hunk 1 above)
Comment (SeyfSV) [reply .github/workflows/install-tests.yml:6]: ok

File: .github/workflows/unit-tests.yml (hunk 4 above)
Comment (SeyfSV) [reply .github/workflows/unit-tests.yml:18]: There are no reasons not running. But there are issues with running :)
Windows:
* current github windows runner does not allow to run non-windows containers:
image
* ibmcom/mq repository does not have windows version
MacOS:
* [runner](https://github.com/actions/virtual-environments/blob/master/images/macos/macos-10.15-Readme.md) does not have docker at all.
No, these comments should be here, to answering on similar questions. (imho)

File: .github/workflows/unit-tests.yml (hunk 5 above)
Comment (SeyfSV) [reply .github/workflows/unit-tests.yml:120]: It come from action, it is output [variable](https://github.com/SeyfSV/setup-mqclient/blob/766fcab5f19b061c72fbc9d65e875397492df89d/main.js#L135):
And this is its [initialization](https://github.com/SeyfSV/setup-mqclient/blob/766fcab5f19b061c72fbc9d65e875397492df89d/main.js#L39-L63):
```
    case "linux":
        url = REDIST_URL_LNX
        archive_name = ARCHIVE_LNX
        if (mq_file_path == ''){
            core.debug(`$HOME variable is ${process.env.HOME}`)
            mq_file_path = path.join(process.env.HOME,'IBM/MQ/data')
        }
        break;
    case "win32":
        url = REDIST_URL_WIN
        archive_name = ARCHIVE_WIN
        if (mq_file_path == ''){
            core.debug(`%HOMEDRIVE% variable is ${process.env.HOMEDRIVE}`)
            core.debug(`%HOMEPATH% variable is ${process.env.HOMEPATH}`)
            mq_file_path = path.join(process.env.HOMEDRIVE, process.env.HOMEPATH, 'IBM/MQ/data')
        }
        break;
    case "darwin":
        url = TOOLKIT_URL_MAC
        archive_name = ARCHIVE_MACcore.setOutput('mq-file-path', `${mq_file_path}`)
        if (mq_file_path == ''){
            core.debug(`$HOME variable is ${process.env.HOME}`)
            mq_file_path = path.join(process.env.HOME,'IBM/MQ/data')
        }
        break;
```

**Comment by dsuch** [comment]
@AlexandreYang - I am not sure what you meant by "; [tox]" but whatever you do, can you please make sure that tox.ini is really as here in this file?

https://github.com/dsuch/pymqi/blob/master/tox.ini

This is a set of contemporary settings which are not strictly PEP-8 and that is a good thing because I have found PEP-8 much too restrictive.

File: code/README (hunk 6 above)
Comment (SeyfSV) [reply code/README:2]: No. We can, but... There is no differences as I know.

File: code/tests/test_rfh2.py (hunk 8 above)
Comment (SeyfSV) [reply code/tests/test_rfh2.py]: By mistake.
Fixed 04ec551fc1ed24252240cc93926be56b091e0c15

File: code/tests/test_pcf.py (hunk 9 above)
Comment (SeyfSV) [reply code/tests/test_pcf.py:276]: Is it OK if I change description to this [link](https://github.com/dsuch/pymqi/issues/207#issuecomment-645422229) ?
No, it should be skipped only for pre 3.7. It was quick fixpack for #207 issue, it still open and will be changed.

File: code/tests/test_pcf.py (hunk 10 above)
Comment (SeyfSV) [reply code/tests/test_pcf.py:26]: Yes, thanks.
Fixed 8afecb287ce52a688d951377e408f0185f8fc976

File: code/tests/test_queue_manager.py (hunk 11 above)
Comment (SeyfSV) [reply code/tests/test_queue_manager.py:154]: I think it fully covered by others, for example, `test_disconnect` from `test_queue_manager.py`

**Comment by SeyfSV** [comment]
@dsuch `; [tox]` is just a rubbish from first try to create multi-version testing.
It was something like
```
[tox]
envlist = py27,py36
```

File: .github/workflows/install-tests.yml (hunk 2 above)
Comment (SeyfSV) [reply .github/workflows/install-tests.yml]: 2.7:
Windows runner does not have libraries required for build with python2.7 
image
And I couldn't find a way to install it, it was terrible :(

3.5:
Logs is gone, and I could not show what was wrong. But it did not work. Now it's OK: 74c2aefc577b1d0f1a4ec614dc427dbd8f86914c

File: code/tests/test_pcf.py (hunk 9 above)
Comment (AlexandreYang) [reply code/tests/test_pcf.py:274]: I don't have enough context on this, but yes, providing a comment link will help.

File: .github/workflows/install-tests.yml (hunk 2 above)
Comment (AlexandreYang) [reply .github/workflows/install-tests.yml]: Thx for the explanation.

Adding a comment like "Windows runner does not have libraries required for build with python2.7" might help I think.

File: .github/workflows/unit-tests.yml (hunk 4 above)
Comment (AlexandreYang) [reply .github/workflows/unit-tests.yml:18]: Looks good then. Thx for the explanation.

File: .github/workflows/unit-tests.yml (hunk 5 above)
Comment (AlexandreYang) [reply .github/workflows/unit-tests.yml:120]: Oh, I see, thx.

File: code/README (hunk 6 above)
Comment (AlexandreYang) [reply code/README:2]: It's more for convenience.

Without .md

image


With .md

image


.md extension will also help IDEs to recognise the type of file.

File: code/tests/test_queue_manager.py (hunk 11 above)
Comment (AlexandreYang) [reply code/tests/test_queue_manager.py:154]: Thx for the explanation.

**Comment by AlexandreYang** [comment]
> @AlexandreYang - I am not sure what you meant by "; [tox]" but whatever you do, can you please make sure that tox.ini is really as here in this file?

@dsuch my comment was a "GitHub review suggestion" to remove the commented "; [tox]"

File: .github/workflows/install-tests.yml (hunk 2 above)
Comment (SeyfSV) [reply .github/workflows/install-tests.yml]: Fixed 6f2d52f055eff821b0ac86409041b3ed6c0c697b

File: code/tests/test_pcf.py (hunk 9 above)
Comment (SeyfSV) [reply code/tests/test_pcf.py]: Fixed 6f2d52f055eff821b0ac86409041b3ed6c0c697b
* Link added
* Test skipped only for pre 3.7

**Comment by SeyfSV** [comment]
@dsuch setup.py was moved to root dir next to tox.ini file, as stated in tox [documentation](https://tox.readthedocs.io/en/latest/example/basic.html):
> ...into a tox.ini file that should reside next to your setup.py file

**Pull request merged**